### PR TITLE
fix: reclaim orphaned worktree/session state on load (duplicate workspace that returns after restart)

### PR DESCRIPTION
**Pull request by PannenetsF**
## Summary

A project can appear **twice** in the sidebar — one real card plus a phantom duplicate at the *same path* — and the phantom **returns after every restart**, even after you delete it from `worktreeMeta` by hand. This is the already-orphaned sibling of #9024: the repo id is gone from `repos`, but its `worktreeMeta` row and `workspaceSession` owner keys survive forever and get re-materialized on load.

This is the **load-side companion to #9025**. #9025 fixes the *forward* path — deleting a project that is still in `repos`. It does **not** help a repo id that is already gone from `repos`, because the function it patches is unreachable in that state.

### Root cause

Two gaps combine so an ownerless entry becomes permanently unreclaimable:

1. **`pruneWorktreeStateForRepo` is unreachable once the repo leaves `repos`.** Its only callers are `removeProject` and `removeProjectForHost` (`persistence.ts`), and every external entry point into those resolves a *live* repo first (`ipc/repos.ts`, `orca-runtime.ts`). So a repo id that left `repos` by any other path — a partial/interrupted delete, a repo-id re-key/migration (`renameWorktreeId`), or a crash between the `repos` filter and the save — is never routed through prune again. #9025's added session-owner cleanup never runs for such an id.

2. **`gcStaleWorktreeMeta` deliberately keeps the entry.** The only automatic worktreeMeta cleanup skips an unowned key when the meta carries a non-local `hostId`, and — critically — when `existsSync(worktreePath)` is true. For a *duplicate that shares a live repo's path*, `existsSync` is always true (the surviving repo occupies that path), so the orphan is never collected. It is then restored verbatim on every load (the load path applies no repo-membership filter to `worktreeMeta`).

The **visible** duplicate comes from session state, which prune never touched: `hydrateHeadlessMobileSessionTabsFromWorkspaceSession` iterates `workspaceSession.tabsByWorktree` with **no repo gate**, so a leftover `${repoId}::${path}` owner key re-creates a live workspace on boot — with no `session-created` event. That is the second card.

### Fix

Add `gcOrphanedRepoState`, a load-time sweep (runs right after `gcStaleWorktreeMeta`) that enforces the missing invariant: **worktree and session state must not outlive its owning repo/project id.** For every `worktreeMeta` / `worktreeLineageById` / `workspaceLineageByChildKey` key and every `workspaceSession` (+ `workspaceSessionsByHostId`) owner key whose parsed repo id is absent from **both** `repos` and `projects`, it removes the entry — reusing the session-owner cleanup so the sweep and the delete path share one implementation. Unlike `gcStaleWorktreeMeta` it does not apply the host/`existsSync`/grace guards, because a truly ownerless key can never be legitimately displayed (enumeration is repo-driven) regardless of host or path existence.

**Safety guard:** the sweep is skipped when there are zero live owners, so an anomalous empty-`repos` load can never wipe all worktree/session state.

This PR also refactors `removeWorkspaceSessionOwner` into a reusable `removeWorkspaceSessionOwnerInPlace` / `removeWorkspaceSessionOwners` pair (and now also clears `remoteSessionIdsByTabId` for removed tabs), so the single-owner delete path and the multi-owner sweep share one cleanup routine.

## Screenshots

No visual change. The change is entirely in the main-process persistence layer (`src/main/persistence.ts`); no renderer UI is touched. The user-observable effect is behavioral: a phantom duplicate project card that previously reappeared after every restart is now removed on load.

## Testing

- [ ] `pnpm lint`
- [ ] `pnpm typecheck`
- [ ] `pnpm test`
- [ ] `pnpm build`
- [x] Added or updated high-quality tests for my change (or explained below why not)

New `persistence.test.ts` coverage (author-described):
- orphan `worktreeMeta` sharing a **live repo's path** is swept (the case `gcStaleWorktreeMeta` keeps because `existsSync` is true), as is one carrying a **remote `hostId`**;
- `workspaceSession` owner keys are swept across the **legacy blob and a host partition** (tabs, layouts, `lastVisitedAtByWorktreeId`, `activeWorktreeId`, `activeWorktreeIdsOnShutdown`, `remoteSessionIdsByTabId`, `openFilesByWorktree`), leaving the live repo's session intact;
- the **empty-owner gate** keeps everything when `repos` is empty;
- a lineage-only orphan is swept and the load is marked for persistence.

Existing `gcStaleWorktreeMeta` tests are unaffected (a repo still present on any host stays "owned" and is never swept). Note: `pnpm` checks were not run in this environment and the boxes above are left unchecked accordingly.

## AI Review Report

Reviewed the full diff. It is confined to `src/main/persistence.ts` (+149/-8) and its test file (+131/-8).

Correctness / edge cases examined:
- **Owner-id parsing.** `ownerIdOf` returns `null` for folder-scoped keys (correctly excluding them from the repo-membership check) and slices the id before the first `::`. Keys without a `::` separator yield `null` and are treated as non-orphan (kept), which is conservative and safe.
- **Empty-owner guard.** The early return when `liveOwnerIds.size === 0` is the key safety property — it prevents a corrupt/empty-`repos` load from deleting all worktree and session state. Covered by a dedicated test.
- **Lineage sweeps.** Both `worktreeLineageById` (checks child id and `parentWorktreeId`) and `workspaceLineageByChildKey` (checks parsed child and parent worktree scopes) are handled; a lineage-only orphan with no meta/session twin is exercised by a test that also asserts the change is persisted.
- **Session cleanup reuse.** Refactoring single-owner removal into an in-place helper reused by the batch sweep reduces divergence risk. The added `remoteSessionIdsByTabId` cleanup closes a real leak (tab-scoped remote session ids previously outlived removed tabs).
- **`activeRepoId` reset.** Reset to `null` when the active repo id is not a live owner, per session partition — consistent with the sweep's intent.
- Residual consideration: the sweep depends on `WORKSPACE_SESSION_OWNER_RECORD_KEYS` staying in sync with the `WorkspaceSessionState` shape; the `satisfies readonly (keyof WorkspaceSessionState)[]` annotation gives compile-time protection against typos, though a newly-added owner-keyed record would still need to be appended manually.

**Cross-platform (macOS / Linux / Windows):** checked. No keyboard shortcuts, shortcut labels, shell invocation, or Electron main/renderer boundary behavior are touched. Path handling is limited to string parsing of the existing `${repoId}::${path}` key format (splitting on `::`), not filesystem path construction, so it is platform-agnostic; test fixtures use Linux-style paths but exercise pure string logic. No platform-specific code paths were introduced or modified.

## Security Audit

Reviewed for input handling, command execution, path handling, auth, secrets, dependency, and IPC changes.

- **No command execution, no shell, no network, no new dependencies.**
- **No IPC surface change** — the sweep runs internally during store load; no new handlers or renderer-exposed APIs.
- **Path handling** is pure in-memory string parsing of persisted keys, not filesystem access (no new `existsSync`/`fs` calls; the sweep intentionally omits the `existsSync` guard). No path traversal or injection surface.
- **Data-loss risk** is the only material concern and is directly mitigated by the empty-owner guard, which prevents wiping all state on an anomalous load; a targeted test asserts this. The sweep only deletes state whose owning repo/project id is provably absent from both `repos` and `projects`.
- No secrets, credentials, or auth logic touched.

No security issues found.

## Notes

- Persistence-layer only; effect is observed on next app load (orphaned state is swept and the load is flagged for save).
- SSH / remote hosts are covered: the sweep iterates `workspaceSessionsByHostId` partitions and intentionally sweeps orphan keys regardless of `hostId` (including remote), since an ownerless key is never legitimately enumerable.
- Optional follow-up (not in this PR): gate `hydrateHeadlessMobileSessionTabsFromWorkspaceSession` on repo membership as defense-in-depth. The load-time sweep already removes the orphan before that runtime consumer sees it, so it is not required for the fix.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

## ELI5

A deleted project could leave a phantom duplicate workspace that came back after every restart. On load, Orca reclaims orphaned worktree and session rows so ghosts do not reappear.

### CODE REVIEW

File: src/main/persistence.ts [hunk 1]
```diff
@@ -435,6 +435,103 @@ function gcStaleWorktreeMeta(state: PersistedState): number {
...
+  for (const [childKey, lineage] of Object.entries(state.workspaceLineageByChildKey)) {
+    const childScope = parseWorkspaceKey(childKey)
+    const parentScope = parseWorkspaceKey(lineage.parentWorkspaceKey)
+    if (
+      (childScope?.type === 'worktree' && isOrphanKey(childScope.worktreeId)) ||
+      (parentScope?.type === 'worktree' && isOrphanKey(parentScope.worktreeId))
+    ) {
+      delete state.workspaceLineageByChildKey[childKey as WorkspaceKey]
+    }
+  }
```
[the remark this answers is not in the document]
Comment (PannenetsF) [reply src/main/persistence.ts:464]: Good catch — fixed in `a67cafcf9`. `removed++` now fires in both the `worktreeLineageById` and `workspaceLineageByChildKey` deletion blocks, so a profile whose only orphaned state is a lineage entry sets `loadNeedsSave` and the sweep persists. Added a regression test (`sweeps and persists a lineage entry whose parent repo id is gone`).

**Comment by brennanb2025** [comment]
Taking a look!

**Comment by brennanb2025** [comment]
Thanks for the thorough report and regression coverage! I rebased the branch onto current main and tightened the load-time reconciliation so repos/projects remain authoritative across worktree metadata, lineage, all owner-keyed session fields, and host partitions. The follow-up preserves the lineage dirty-count fix, handles canonical workspace keys and dependent relay state, and batches cleanup to one clone per session instead of one per orphan key. Coverage now includes editor-only orphan state, canonical active keys, relay IDs, host partitions, live/remote paths, empty-owner safety, and lineage-only persistence; all 386 persistence tests, node typecheck, touched-file lint, and the max-lines ratchet pass.

File: src/main/persistence.ts [hunk 2]
```diff
@@ -415,6 +415,71 @@ function gcStaleWorktreeMeta(state: PersistedState): number {
   return removed
 }
 
+function gcOrphanedRepoState(state: PersistedState): number {
+  const liveOwnerIds = new Set([
+    ...state.repos.map((repo) => repo.id),
+    ...(state.projects ?? []).map((project) => project.id)
+  ])
```
[the remark this answers is not in the document]
Comment (PannenetsF) [reply src/main/persistence.ts]: Fixed in 7091a01. Orphan detection is now scoped to `(repoId, executionHostId)`: I build a `liveRepoHostKeys` set from `getRepoExecutionHostId(repo)` and check host-tagged `worktreeMeta` (by `meta.hostId`) and each `workspaceSessionsByHostId` partition (by its host key) against it, so a live local `repo1` no longer keeps a removed SSH `repo1`'s meta/partition alive. The legacy `workspaceSession` blob and lineage keep the conservative any-host check, since they can hold a remote repo's owners and carry no host of their own. Added a duplicate-id local/SSH regression test.

File: src/main/persistence.ts [hunk 3]
```diff
@@ -2370,6 +2435,44 @@ function cloneWorkspaceSessionState(session: WorkspaceSessionState): WorkspaceSe
...
+  if (session.activeWorktreeId) {
+    ownerKeys.add(session.activeWorktreeId)
+  }
+  for (const worktreeId of session.activeWorktreeIdsOnShutdown ?? []) {
+    ownerKeys.add(worktreeId)
+  }
+  for (const record of Object.values(session.sleepingAgentSessionsByPaneKey ?? {})) {
+    ownerKeys.add(record.worktreeId)
+  }
+  return ownerKeys
```
[the remark this answers is not in the document]
Comment (PannenetsF) [reply src/main/persistence.ts]: Fixed in 7091a01. `collectWorkspaceSessionOwnerKeys` now also iterates `terminalSurfaceTombstonesByPaneKey` and adds each `tombstone.worktreeId`, so a session that retains only a tombstone for an orphaned worktree produces an owner key and gets swept. Added a tombstone-only regression test.

**Comment by PannenetsF** [comment]
@brennanb2025 Rebased onto the latest main and resolved all conflicts. Verified locally (typecheck + targeted tests pass); CI is pending maintainer approval to run. Ready for review whenever you have a moment — thanks!
